### PR TITLE
feat: Linear Attention Dispatch

### LINKED ISSUE #947: [Feature] Support HybridLinearAttnBackend
### Checklist

- [ ] 1. If the issue you raised is not a feature but a question, please raise a discussion at https://github.com/sgl-project/sgl-jax/discussions/new/choose Otherwise, it will be closed.
- [ ] 2. Please use English, otherwise it will be closed.

### Motivation

Kimi-Linear consists of KDA and MLA which is a hybrid architecture. In order to the upstream like `model_runner` be unperceived to this modification, we have to support this architecture.

### Related resources

_No response_

### PR DESCRIPTION
**Pull request by MokusMokun**
## Motivation

Adds infrastructure for hybrid linear-attention models (e.g., Kimi-Linear) that mix softmax attention layers with linear recurrent attention layers (KDA, GLA, Mamba2). This PR delivers the dispatch framework; KDA kernel implementation follows in a separate PR.

Depends on #1034 (Hybrid Memory Pool).

### 1. `HybridLinearAttnBackend` — per-layer dispatch
- Routes by `layer.layer_id` to either:
  - Full attention sub-backend (MLA / FlashAttention)
  - Linear attention sub-backend (KDA / GLA / Mamba2)
- Aggregates metadata from both sub-backends
- Integrated via `attn_backend_wrapper` factory in `ModelRunner`

### 2. `LinearRecurrentAttnBackend` — base class for linear attention
- Provides `get_forward_metadata()` with DP-aware sharding:
  - `cu_q_lens`: per-rank cumulative sum boundaries, sharded `P("data")`
  - `recurrent_indices`: pool slot indices, sharded `P("data")`
  - `has_initial_state`: new vs continuing requests, sharded `P("data")`
- Follows FlashAttention's 2D cumsum pattern for DP metadata construction
- Handles single-process (DP=1) and multi-process (DP>1) correctly

### 3. `RadixLinearAttention` — layer entry point
- nnx.Module analogous to `RadixAttention`
- Holds conv1d weight references, gate parameters (`A_log`, `dt_bias`), scale
- Delegates to backend via `forward_batch.attn_backend()`

### 4. `LinearRecurrentAttnBackendMetadata`
- Pytree-registered metadata container
- Transfers `has_initial_state` from CPU (`ModelWorkerBatch`) to device with proper sharding

## Files Changed

**New:**
- `srt/layers/attention/hybrid_linear_attn_backend.py` — dispatch framework
- `srt/layers/radix_linear_attention.py` — layer entry point

## Test Plan

Tested via KDA implementation PR (follows this one). This PR provides the framework only; no standalone tests.

## Checklist

- [x] English only
- [x] Purpose: enable hybrid linear-attention model serving
- [x] Depends on #1034
- [x] Test plan: covered by downstream KDA PR

### CODE REVIEW

File: python/sgl_jax/srt/layers/attention/hybrid_linear_attn_backend.py [hunk 1]
```diff
@@ -0,0 +1,277 @@
...
+        ) = device_array(
+            (cu_q_lens, batch.recurrent_indices, batch.has_initial_state),
+            sharding=(
+                (
+                    NamedSharding(self.mesh, P("data")),  # cu_q_lens
+                    NamedSharding(self.mesh, P("data")),  # recurrent_indices
+                    NamedSharding(self.mesh, P("data")),  # has_initial_state
+                )
+                if self.mesh is not None and jax.process_count() > 1
+                else None
```
Comment (MokusMokun) [inline python/sgl_jax/srt/layers/attention/hybrid_linear_attn_backend.py]: this if check might be redundant, just
`sharding=(NamedSharding(self.mesh, P("data"))),` is enough?

Comment (JamesBrianD) [reply python/sgl_jax/srt/layers/attention/hybrid_linear_attn_backend.py]: Yes, `sharding=(NamedSharding(self.mesh, P("data")))` is enough. And not need to retain the 'if' condition.

Comment (aolemila) [reply python/sgl_jax/srt/layers/attention/hybrid_linear_attn_backend.py]: I think yes. Current main has removed this condition. cc @JamesBrianD

File: python/sgl_jax/srt/layers/attention/hybrid_linear_attn_backend.py [hunk 2]
```diff
@@ -0,0 +1,277 @@
...
+    def tree_unflatten(cls, aux_data, children):
+        obj = cls()
+        obj.forward_metadata = children[0]
+        return obj
+
+    @staticmethod
+    def get_layer_cache(recurrent_state_pool, layer_id: int):
+        """Returns (recurrent_cache, conv_cache) for the given layer.
+
+        Matches RecurrentStatePool.get_linear_recurrent_layer_cache (PR #966)
```
Comment (JamesBrianD) [inline python/sgl_jax/srt/layers/attention/hybrid_linear_attn_backend.py]: Clean up the comments here.

File: python/sgl_jax/srt/layers/attention/hybrid_linear_attn_backend.py [hunk 3]
```diff
@@ -0,0 +1,277 @@
...
+            raise ImportError(
+                "HybridLinearAttnBackend needs KDAAttnBackend " "(delivered by a separate PR)."
+            ) from e
+
+        linear_attn_backend = KDAAttnBackend(runner.mesh)
+    else:
+        raise NotImplementedError(f"No linear backend wired for hybrid config {type(cfg).__name__}")
+    return HybridLinearAttnBackend(
+        full_attn_backend, linear_attn_backend, cfg.full_attention_layer_ids
+    )
```
Comment (JamesBrianD) [inline python/sgl_jax/srt/layers/attention/hybrid_linear_attn_backend.py]: Currently, there are no implementations for the Kimi linear config or backend; therefore, remove the redundant code and retain only the `return full_attn_backend` statement.

File: python/sgl_jax/srt/layers/attention/hybrid_linear_attn_backend.py [hunk 4]
```diff
@@ -0,0 +1,277 @@
+"""HybridLinearAttnBackend — dispatches per-layer to a full-attention sub-backend
+(MLA / FlashAttention) or a linear-attention sub-backend (KDA).
+
+The class itself owns no memory pool and allocates no device buffers; it only
+holds two sub-backends + a `full_attn_layers` whitelist and routes calls.
+"""
```
Comment (JamesBrianD) [inline python/sgl_jax/srt/layers/attention/hybrid_linear_attn_backend.py]: The comments can be removed.

File: python/sgl_jax/srt/layers/attention/hybrid_linear_attn_backend.py [hunk 5]
```diff
@@ -0,0 +1,277 @@
...
+            b=b,
+            recurrent_state_pool=pool,
+            **kwargs,
+        )
+
+    def get_max_running_reqests(self, max_context_len, page_size):
+        return self.full_attn_backend.get_max_running_reqests(max_context_len, page_size)
+
+
+def attn_backend_wrapper(
```
Comment (aolemila) [inline python/sgl_jax/srt/layers/attention/hybrid_linear_attn_backend.py:229]: Please add logic in `_get_attention_backend`. You can refer to [epic](https://github.com/sgl-project/sglang-jax/pull/985/changes) 

Image

File: python/sgl_jax/srt/layers/attention/hybrid_linear_attn_backend.py (hunk 3 above)
Comment (MokusMokun) [reply python/sgl_jax/srt/layers/attention/hybrid_linear_attn_backend.py]: removed, to be recovered in future KDA PR

File: python/sgl_jax/srt/layers/attention/hybrid_linear_attn_backend.py (hunk 5 above)
Comment (MokusMokun) [reply python/sgl_jax/srt/layers/attention/hybrid_linear_attn_backend.py:229]: `model_runner.py: _get_attention_backend` has been refactored

**Review by aolemila — APPROVED** [review]

**Comment by MokusMokun** [comment]
only one performance CI failed, ready to merge ?
@aolemila @JamesBrianD

**Comment by JamesBrianD** [comment]
> only one performance CI failed, ready to merge ? @aolemila @JamesBrianD

Merged, flaky ci failed.
